packages/cross_file/README.md

```diff
@@ -4,15 +4,38 @@ An abstraction to allow working with files across multiple platforms.
 
 ## Usage
 
-Import `package:cross_file/cross_file.dart`, instantiate a `XFile`
-using a path or byte array and use its methods and properties to
-access the file and its metadata.
+Many packages use `XFile` as their return type. In order for your
+application to consume those files, import 
+`package:cross_file/cross_file.dart`, and use its methods and properties
+to access the file data and metadata.
 
-Example:
+In order to instantiate a new `XFile`, import the correct factory class,
+either from `package:cross_file/native/factory.dart` (for native development) or
```

Comment:
Small drive-by comment:

Since the two implementations now have platform-specific imports (dart:io or package:web)
users who take the imports at face value might run into compilation issues, because they lack a conditional import?

Can we provide guidance on how to avoid this? With the legacy API we had the File or Blob types as internals and not on the public interface, which allowed for the plugin to provide a conditional import shim.

Reply:
@navaronbracke thanks for the comment! This is a very interesting problem, for which I don't have a great answer. The TL;DR/Guidance is:

* Are you reading XFiles? Use the `XFile` interface only.
* Are you **creating** XFiles?
  * In mobile? Use the `native/factory.dart`
  * In web? Use the `web/factory.dart`

Each factory exposes constructors that are platform-specific, and we really don't have a cross-platform file constructor.

What problems do you think people are going to have? Is this a documentation problem, or a design issue?

Reply:
If people take the comment at face value and do `import "package:cross_file/web/factory.dart";` directly, they will run into errors like `library dart:js_interop is not available on this platform`. Since they don't directly import `dart:js_interop`, this might be confusing to new users.

I think this is merely a documentation issue. We should put emphasis on:
1) If this is in a platform-specific plugin implementation, you can probably import it directly (as the plugin system only uses the platform implementations of your plugin on the correct platforms)
2) If this is in a Flutter app, you'll have to use a conditional import on `if (dart.library.js_interop) ...` and you probably need an extra abstraction, since the web-only types (i.e. Blob) cannot be imported in native implementations (for the same reason why the plugin cannot provide the conditional import here)

Reply:
OK, if it's a documentation issue, it's probably easier to fix (LOL famous last words :P)

Do you have any ideas with how this comment would be clearer? What would you like to read?

(I'll try to come up with something more specific as well!)

Reply:
Hmm, the main point is that the compiler itself would want you to use a conditional import.
So perhaps something like: 

"In order to instantiate a new `XFile`, import the correct factory class, through a conditional import.
Using a conditional import is required to use types that are only available on the current platform.

```
import 'package:cross_file/native/factory.dart'
 if (dart.library.js_interop) 'package:cross_file/web/factory.dart'
```
"

The wording is just a suggestion, so feel free to tweak it.

Reply:
Hmm, wait. Since the API for both factories is divergent, would a user that uses this conditional import know at compile time if they could access the `fromFile()` (both with a dart:io or package:web File) or the `fromBlob()/fromObjectUrl()` methods? I think that Dart will only provide the base type `XFileFactory`, but I _think_ you don't get a guarantee for the available methods.

Perhaps we can fix that by having the interface like:

- `XFileFactory` is the base type that is imported through a conditional import. It provides all the methods that do not include platform-specific types (so only fromPath/fromBytes/fromObjectUrl)
- `WebXFileFactory` and `NativeXFileFactory` implement that interface
- `WebXFileFactory` provides the `fromBlob/fromFile` implementations for package:web types
- `NativeXFileFactory` provides the `fromFile` implementation for dart:io
- unsupported methods on a platform just throw `UnsupportedError`
- we direct users to use `kIsWeb` and/or a type check on the factory

Reply:
Thanks for the clarification @navaronbracke! I see what you mean! You want the `XFileFactory` to be cross-platform as well!

IMO the only cross-platform class from this package is **the `XFile` interface**! [^1]

If your cross-platform App needs to "create XFiles", you'll end up with some platform-aware code (because as you said, you can't import web stuff in native code), like this:

```dart
// Your app's main.dart
import 'src/get_file/get_file.dart'; // Platform-aware, see below

// ...

final XFile myFile = await getFile(/* config? data? */);
```

With a directory like this:

```
src/
  get_file/
    get_file.dart
    get_file_web.dart
    get_file_native.dart
```

Your `get_file.dart` could [conditionally export](https://dart.dev/guides/libraries/create-packages#conditionally-importing-and-exporting-library-files) the right file for the platform; like:

```dart
// get_file.dart

export "get_file_native.dart"
  if (dart.library.js_interop) "get_file_web.dart";
```

So each implementation file can use the platform-specific factories:

<table>
<tr>
  <th><tt>get_file_native.dart</tt></th>
  <th><tt>get_file_web.dart</tt></th>
</tr>
<tr>
<td>

```dart
import 'package:cross_file/native/factory.dart';

XFile getFile(/* config? data? */) async {
  // Do the right thing for native here, maybe
  // we write the data to a File in a temp dir
  // and then:
  return XFileFactory.fromFile(tmpFile);
}
```

</td>
<td>

```dart
import 'package:cross_file/web/factory.dart';

XFile getFile(/* config? data? */) async {
  // Do the right thing for the web here,
  // maybe we grab bytes, put them into a
  // Blob and do:
  return XFileFactory.fromBlob(dataBlob);
}
```

</td>
</tr>
</table>

Maybe the confusing part is calling the `XFileFactory` the same on both `native` and `web`? Should we just have `NativeXFileFactory` and `WebXFileFactory` to make the separation clearer, and signal that we never expect a shared interface across both classes?

---

> * It provides all the methods that do not include platform-specific types (so only fromPath/fromBytes/fromObjectUrl)

This is why I think having a base class for the `XFileFactory` is more trouble than what it's worth. Unless I'm missing something, for this approach to work, the `XFileFactory` needs to be an actual `instance` that must be replaced at run-time (similar to a federated plugin?).

Since we can't expose web-only types through this common API, users would still need to **conditionally import the native/web types to cast the instance** to the one that has the platform specific methods. And if you're conditionally importing, just **import the right `XFileFactory` static class?** 😛

[^1]: This is similar to the `Client` interface from `package:http` that has [multiple platform-specific implementations](https://pub.dev/packages/http#choosing-an-implementation) coming from different packages